camel/agents/chat_agent.py

```diff
@@ -485,220 +570,217 @@ def step(
                         msgs=output_messages,
                         terminated=self.terminated,
                         info=info,
                     )
 
-                # Normal function calling
-                tool_call_records.append(
-                    self._step_tool_call_and_update(response)
-                )
-
-            if (
-                response_format is not None
-                and self.model_type.support_native_tool_calling
-            ):
-                (
-                    output_messages,
-                    finish_reasons,
-                    usage_dict,
-                    response_id,
-                    tool_call,
-                    num_tokens,
-                ) = self._structure_output_with_function(response_format)
-                tool_call_records.append(tool_call)
+            # Single-step mode ends after one iteration
+            if not multi_step:
+                break
 
-            info = self._step_get_info(
+        # Optional structured output
+        if response_format:
+            (
                 output_messages,
                 finish_reasons,
                 usage_dict,
                 response_id,
-                tool_call_records,
+                tool_call,
                 num_tokens,
-            )
+            ) = self._structure_output_with_function(response_format)
+            tool_call_records.append(tool_call)
 
-            if len(output_messages) == 1:
-                # Auto record if the output result is a single message
-                self.record_message(output_messages[0])
-            else:
-                logger.warning(
-                    "Multiple messages returned in `step()`, message won't be "
-                    "recorded automatically. Please call `record_message()` "
-                    "to record the selected message manually."
+        # Final info and response
+        info = self._step_get_info(
+            output_messages,
+            finish_reasons,
+            usage_dict,
+            response_id,
+            tool_call_records,
+            num_tokens,
+            external_tool_request,
+        )
+        self._log_final_output(output_messages)
+        return ChatAgentResponse(
+            msgs=output_messages, terminated=self.terminated, info=info
+        )
+
+    def _extract_tool_call(
+        self, response: Any
+    ) -> Optional[ChatCompletionMessageToolCall]:
+        r"""Extract the tool call from the model response, if present.
+
+        Args:
+            response (Any): The model's response object.
+
+        Returns:
+            Optional[ChatCompletionMessageToolCall]: The parsed tool call if
+                present, otherwise None.
+        """
+        # Check if the response contains tool calls
+        if (
+            self.is_tools_added()
+            and not self.model_type.support_native_tool_calling
+            and "</function>" in response.choices[0].message.content
```

Comment:
I thought we were assuming an OpenAI-like LLM interface for our inference code? Maybe instead of supporting llama tool cooling in ChatAgents and other high level functionalities, we should convert them before they touch any business logic? Consideration also applies for the tool prompt stuff, which might better be done in translation at the model specific code, so that agents can not need to worry about these things (particularly in mind are things extending ChatAgent)

Reply:
Thanks, @CaelumF ! This is a great idea. Moving the tool-calling setting to the model level is indeed more reasonable. How about we plan the further refactor for later? I’ll create a new issue for this, as it could involve significant changes.

Reply:
Yeah that sounds good, this is definitely an improvement still.